src/expr/core/src/sig/mod.rs

```diff
@@ -461,3 +463,65 @@ pub enum FuncBuilder {
 /// A static distributed slice of functions defined by `#[function]`.
 #[linkme::distributed_slice]
 pub static FUNCTIONS: [fn() -> FuncSign];
+
+#[linkme::distributed_slice]
+pub static UDF_RUNTIMES: [UdfRuntimeDescriptor];
+
+pub struct UdfRuntimeDescriptor {
+    /// Returns if the function match this descriptor.
+    pub match_: fn(language: &str, runtime: Option<&str>, link: Option<&str>) -> bool,
+    /// Creates function from options.
+    pub create: fn(opts: CreateFunctionOptions<'_>) -> anyhow::Result<CreateFunctionOutput>,
+    /// Builds UDF runtime from verified options.
+    pub build: fn(opts: UdfOptions<'_>) -> anyhow::Result<Box<dyn UdfRuntime>>,
+}
+
+pub struct CreateFunctionOptions<'a> {
+    pub name: &'a str,
+    pub arg_names: &'a [String],
+    pub arg_types: &'a [DataType],
+    pub return_type: &'a DataType,
+    pub is_table_function: bool,
+    pub as_: Option<&'a str>,
+    pub using_link: Option<&'a str>,
+    pub using_base64_decoded: Option<&'a [u8]>,
+}
+
+pub struct CreateFunctionOutput {
+    pub identifier: String,
+    pub body: Option<String>,
+    pub compressed_binary: Option<Vec<u8>>,
+}
+
+pub struct UdfOptions<'a> {
+    pub table_function: bool,
+    pub body: Option<&'a str>,
+    pub compressed_binary: Option<&'a [u8]>,
+    pub link: Option<&'a str>,
+    pub identifier: &'a str,
+    pub arg_names: &'a [String],
+    pub return_type: &'a DataType,
+    pub always_retry_on_network_error: bool,
+    pub function_type: Option<&'a str>,
+}
+
+#[async_trait::async_trait]
+pub trait UdfRuntime: std::fmt::Debug + Send + Sync {
```

Comment:
I think this should be called `Impl` (which uses a `Runtime` to evaluate) instead of `Runtime`.